File: src/main/java/com/noximity/remmyChat/services/ChatService.java
package com.noximity.remmyChat.services;

import com.noximity.remmyChat.RemmyChat;
import com.noximity.remmyChat.models.ChatUser;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

public class ChatService {

    private final RemmyChat plugin;
    private final Map<UUID, ChatUser> chatUsers = new HashMap<>();

    public ChatService(RemmyChat plugin) {
        this.plugin = plugin;
    }

    public ChatUser getChatUser(UUID uuid) {
        return chatUsers.computeIfAbsent(uuid, id -> {
            String defaultChannel = plugin.getConfigManager().getDefaultChannel().getName();
            return plugin.getDatabaseManager().loadUserPreferences(id, defaultChannel);
        });
    }

    public void createChatUser(UUID uuid) {
        if (!chatUsers.containsKey(uuid)) {
            String defaultChannel = plugin.getConfigManager().getDefaultChannel().getName();
            ChatUser user = plugin.getDatabaseManager().loadUserPreferences(uuid, defaultChannel);
            chatUsers.put(uuid, user);
        }
    }

    public void removeChatUser(UUID uuid) {
        if (chatUsers.containsKey(uuid)) {
            plugin.getDatabaseManager().saveUserPreferences(chatUsers.get(uuid));
            chatUsers.remove(uuid);
        }
    }

    public boolean setChannel(UUID uuid, String channel) {
        if (plugin.getConfigManager().getChannel(channel) == null) {
            return false;
        }

        getChatUser(uuid).setCurrentChannel(channel);
        return true;
    }

    public void saveAllUsers() {
        List<ChatUser> users = new ArrayList<>(chatUsers.values());
        for (ChatUser user : users) {
            plugin.getDatabaseManager().saveUserPreferences(user);
        }
    }

    public List<ChatUser> getSocialSpyUsers() {
        List<ChatUser> spyUsers = new ArrayList<>();
        for (ChatUser user : chatUsers.values()) {
            if (user.isSocialSpy()) {
                spyUsers.add(user);
            }
        }
        return spyUsers;
    }
}

File: bin/main/com/noximity/remmyChat/services/ChatService.kt
package com.noximity.remmyChat.services

import com.noximity.remmyChat.RemmyChat
import com.noximity.remmyChat.models.ChatUser
import java.util.*

class ChatService(private val plugin: RemmyChat) {
    private val chatUsers: MutableMap<UUID?, ChatUser> = HashMap<UUID?, ChatUser>()

    fun getChatUser(uuid: UUID?): ChatUser? {
        return chatUsers.computeIfAbsent(uuid) { id: UUID? ->
            val defaultChannel = plugin.getConfigManager().getDefaultChannel().getName()
            plugin.getDatabaseManager().loadUserPreferences(id, defaultChannel)
        }
    }

    fun createChatUser(uuid: UUID) {
        if (!chatUsers.containsKey(uuid)) {
            val defaultChannel = plugin.getConfigManager().getDefaultChannel().getName()
            val user = plugin.getDatabaseManager().loadUserPreferences(uuid, defaultChannel)
            chatUsers.put(uuid, user!!)
        }
    }

    fun removeChatUser(uuid: UUID?) {
        if (chatUsers.containsKey(uuid)) {
            plugin.getDatabaseManager().saveUserPreferences(chatUsers.get(uuid))
            chatUsers.remove(uuid)
        }
    }

    fun setChannel(uuid: UUID?, channel: String?): Boolean {
        if (plugin.getConfigManager().getChannel(channel) == null) {
            return false
        }

        getChatUser(uuid)!!.setCurrentChannel(channel)
        return true
    }

    fun saveAllUsers() {
        val users: MutableList<ChatUser?> = ArrayList<ChatUser?>(chatUsers.values)
        for (user in users) {
            plugin.getDatabaseManager().saveUserPreferences(user)
        }
    }

    val socialSpyUsers: MutableList<ChatUser?>
        get() {
            val spyUsers: MutableList<ChatUser?> = ArrayList<ChatUser?>()
            for (user in chatUsers.values) {
                if (user.isSocialSpy()) {
                    spyUsers.add(user)
                }
            }
            return spyUsers
        }
}